### PR TITLE
Refactor IoPath FileLister and FileLoader

### LINKED ISSUE #52: Use IoPath to download HTTP Url
Since IoPath has features to download different sources: HTTP, S3, Google, and etc, we should add a `IoPathDownloader` to handle all different sources.

### PR DESCRIPTION
**Pull request by ejguan**
Fixes #52

This PR is going to add all existing `PathHandler`s from `iopath` to make sure it can handle the following data sources:
- Local File Path
- HTTP Url
- OneDrive Url
- S3 Url (will be included in `iopath` next release)

Before this PR, these two DataPipes only work with `NativePathHandler`, which would handle the local file.

### CODE REVIEW

File: torchdata/datapipes/iter/load/iopath.py [hunk 1]
```diff
@@ -1,73 +1,108 @@
...
                 "Package `iopath` is required to be installed to use this "
-                "datapipe. Please use `pip install iopath` or `conda install "
-                "iopath`"
-                "to install the package"
+                "datapipe. Please use `pip install iopath` to install the package"
             )
 
         self.root: str = root
-        self.pathmgr = g_pathmgr
+        self.pathmgr = _create_default_pathmanager()
```
Comment (NivekT) [inline torchdata/datapipes/iter/load/iopath.py:61]: Should we give users the option to pass in custom `PathManager` as an argument with None as the default?

Then if `None`, call `_create_default_pathmanager()`

File: torchdata/datapipes/iter/load/iopath.py [hunk 2]
```diff
@@ -6,30 +6,43 @@
...
-        try:
-            from iopath.common.file_io import g_pathmgr
-        except ImportError:
+        if PathManagerFactory is None:
             raise ModuleNotFoundError(
                 "Package `iopath` is required to be installed to use this "
-                "datapipe. Please use `pip install iopath` or `conda install "
-                "iopath`"
-                "to install the package"
+                "datapipe. Please use `pip install iopath` to install the package"
```
[the remark this answers is not in the document]
Comment (NivekT) [reply torchdata/datapipes/iter/load/iopath.py:57]: I am out of the loop but what is the issue with conda-forge?

File: torchdata/datapipes/iter/load/iopath.py [hunk 3]
```diff
@@ -1,73 +1,108 @@
...
             )
 
         self.root: str = root
-        self.pathmgr = g_pathmgr
+        self.pathmgr = _create_default_pathmanager()
+        self.masks = masks
+
+    def register_handler(self, handler, allow_override=False):
```
Comment (NivekT) [inline torchdata/datapipes/iter/load/iopath.py:64]: Do the existing test cases cover these change? Or do we plan to have a separate PR to write a complete set of test for these DataPipes?

File: torchdata/datapipes/iter/load/iopath.py (hunk 1 above)
Comment (ejguan) [reply torchdata/datapipes/iter/load/iopath.py:61]: Makes sense

**Review by NivekT — COMMENTED** [review]
Have some questions but overall looks good

File: torchdata/datapipes/iter/load/iopath.py (hunk 3 above)
Comment (ejguan) [reply torchdata/datapipes/iter/load/iopath.py:64]: It requires to have remote data sources. Can't really any test for now.

Comment (ejguan) [reply torchdata/datapipes/iter/load/iopath.py:64]: We only have test for local files.

**Comment by ejguan** [comment]
Re-create PR to use ghstack.
